[BUG] Elasticsearch throws exception deserializing LocalDate

**Overview of the issue**
When we try to deserialize an entity that was persisted to elasticsearch and has a member of type LocalDate it throws an exception.

Search for something that returns a empty result (running fine)
```
[DEBUG] com.mycompany.myapp.aop.logging.LoggingAspect - Enter: com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys() with argument[s] = [First]
[ERROR] com.mycompany.myapp.aop.logging.LoggingAspect - Exception in com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys() with cause = com.fasterxml.jackson.databind.JsonMappingException: No suitable constructor found for type [simple type, class java.time.LocalDate]: can not instantiate from JSON object (missing default constructor or creator, or perhaps need to add/enable type information?)
 at [Source: {"id":1,"name":"First","startDate":{"year":2014,"month":"NOVEMBER","dayOfMonth":30,"dayOfWeek":"SUNDAY","era":"CE","dayOfYear":334,"leapYear":false,"monthValue":11,"chronology":{"id":"ISO","calendarType":"iso8601"}}}; line: 1, column: 37] (through reference chain: com.mycompany.myapp.domain.MyEntity["startDate"])
org.springframework.data.elasticsearch.ElasticsearchException: failed to map source [ {"id":1,"name":"First","startDate":{"year":2014,"month":"NOVEMBER","dayOfMonth":30,"dayOfWeek":"SUNDAY","era":"CE","dayOfYear":334,"leapYear":false,"monthValue":11,"chronology":{"id":"ISO","calendarType":"iso8601"}}}] to class MyEntity
	at org.springframework.data.elasticsearch.core.AbstractResultMapper.mapEntity(AbstractResultMapper.java:42) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
	at org.springframework.data.elasticsearch.core.DefaultResultMapper.mapResults(DefaultResultMapper.java:83) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
	at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForPage(ElasticsearchTemplate.java:265) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
	at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForPage(ElasticsearchTemplate.java:259) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
	at org.springframework.data.elasticsearch.repository.support.AbstractElasticsearchRepository.search(AbstractElasticsearchRepository.java:197) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_31]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_31]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.executeMethodOn(RepositoryFactorySupport.java:475) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:460) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:432) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
	at com.sun.proxy.$Proxy186.search(Unknown Source) ~[na:na]
	at com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys(MyEntityResource.java:149) ~[classes/:na]
...
```

**JHipster Version(s)** 
Latest: 2.23.0

Yes, this is a regression. The deserializer is not able to deserialize new Java8 LocalDate.

**JHipster configuration, a `.yo-rc.json` file generated in the root folder** 
Attached as Gist
https://gist.github.com/amitjindal/e15a2124796f3fc5d047

**Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory** 
Attached as Gist
https://gist.github.com/amitjindal/e15a2124796f3fc5d047

**Browsers and Operating System** 
Any

**Reproduce the error** 
Create 2-3 instances of myEntity. Then search for a name that should return one of the rows. On console you will see the exception.

**Related issues** 
None known.

**Suggest a Fix** 
Not sure. Most likely we need to serialize properly or deserialize. Not sure why the default JavaTimeModule does not works. It might be that elasticsearch needs a different registration of this module.

I've tried adding the JsonDeserialize(using = ...)
but that did not make a difference.

Good to see more and more people following our issue guidelines :)
On 28 Oct 2015 22:16, "Amit Jindal" <notifications@github.com> wrote:

> *Overview of the issue*
> When we try to deserialize an entity that was persisted to elasticsearch
> and has a member of type LocalDate it throws an exception.
>
> Search for something that returns a empty result (running fine)
>
> [DEBUG] com.mycompany.myapp.aop.logging.LoggingAspect - Enter: com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys() with argument[s] = [First]
> [ERROR] com.mycompany.myapp.aop.logging.LoggingAspect - Exception in com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys() with cause = com.fasterxml.jackson.databind.JsonMappingException: No suitable constructor found for type [simple type, class java.time.LocalDate]: can not instantiate from JSON object (missing default constructor or creator, or perhaps need to add/enable type information?)
>  at [Source: {"id":1,"name":"First","startDate":{"year":2014,"month":"NOVEMBER","dayOfMonth":30,"dayOfWeek":"SUNDAY","era":"CE","dayOfYear":334,"leapYear":false,"monthValue":11,"chronology":{"id":"ISO","calendarType":"iso8601"}}}; line: 1, column: 37] (through reference chain: com.mycompany.myapp.domain.MyEntity["startDate"])
> org.springframework.data.elasticsearch.ElasticsearchException: failed to map source [ {"id":1,"name":"First","startDate":{"year":2014,"month":"NOVEMBER","dayOfMonth":30,"dayOfWeek":"SUNDAY","era":"CE","dayOfYear":334,"leapYear":false,"monthValue":11,"chronology":{"id":"ISO","calendarType":"iso8601"}}}] to class MyEntity
>     at org.springframework.data.elasticsearch.core.AbstractResultMapper.mapEntity(AbstractResultMapper.java:42) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
>     at org.springframework.data.elasticsearch.core.DefaultResultMapper.mapResults(DefaultResultMapper.java:83) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
>     at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForPage(ElasticsearchTemplate.java:265) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
>     at org.springframework.data.elasticsearch.core.ElasticsearchTemplate.queryForPage(ElasticsearchTemplate.java:259) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
>     at org.springframework.data.elasticsearch.repository.support.AbstractElasticsearchRepository.search(AbstractElasticsearchRepository.java:197) ~[spring-data-elasticsearch-1.3.0.RELEASE.jar:na]
>     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_31]
>     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_31]
>     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
>     at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
>     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.executeMethodOn(RepositoryFactorySupport.java:475) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
>     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:460) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
>     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:432) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
>     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
>     at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
>     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
>     at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
>     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
>     at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207) ~[spring-aop-4.2.2.RELEASE.jar:4.2.2.RELEASE]
>     at com.sun.proxy.$Proxy186.search(Unknown Source) ~[na:na]
>     at com.mycompany.myapp.web.rest.MyEntityResource.searchMyEntitys(MyEntityResource.java:149) ~[classes/:na]
> ...
>
> *JHipster Version(s)*
> Latest: 2.23.0
>
> Yes, this is a regression. The deserializer is not able to deserialize new
> Java8 LocalDate.
>
> *JHipster configuration, a .yo-rc.json file generated in the root folder*
> Attached as Gist
> https://gist.github.com/amitjindal/e15a2124796f3fc5d047
>
> *Entity configuration(s) entityName.json files generated in the .jhipster
> directory*
> Attached as Gist
> https://gist.github.com/amitjindal/e15a2124796f3fc5d047
>
> *Browsers and Operating System*
> Any
>
> *Reproduce the error*
> Create 2-3 instances of myEntity. Then search for a name that should
> return one of the rows. On console you will see the exception.
>
> *Related issues*
> None known.
>
> *Suggest a Fix*
> Not sure. Most likely we need to serialize properly or deserialize. Not
> sure why the default JavaTimeModule does not works. It might be that
> elasticsearch needs a different registration of this module.
>
> I've tried adding the JsonDeserialize(using = ...)
> but that did not make a difference.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/issues/2241>.
>

--- a/app/index.js
+++ b/app/index.js
@@ -725,6 +725,7 @@ JhipsterGenerator.prototype.app = function app() {
     }
 
     if (this.searchEngine == 'elasticsearch') {
+        this.template('src/main/java/package/config/_ElasticSearchConfiguration.java', javaDir + 'config/ElasticSearchConfiguration.java', this, {});
         this.template('src/main/java/package/repository/search/_package-info.java', javaDir + 'repository/search/package-info.java', this, {});
         this.template('src/main/java/package/repository/search/_UserSearchRepository.java', javaDir + 'repository/search/UserSearchRepository.java', this, {});
     }

--- a/app/templates/src/main/java/package/config/_ElasticSearchConfiguration.java
+++ b/app/templates/src/main/java/package/config/_ElasticSearchConfiguration.java
@@ -1,0 +1,44 @@
+package <%=packageName%>.config;
+
+import java.io.IOException;
+
+import org.elasticsearch.client.Client;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.data.elasticsearch.core.ElasticsearchTemplate;
+import org.springframework.data.elasticsearch.core.EntityMapper;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@Configuration
+@AutoConfigureAfter(value = { JacksonConfiguration.class })
+public class ElasticSearchConfiguration {
+
+    @Bean
+    public ElasticsearchTemplate elasticsearchTemplate(Client client, ObjectMapper objectMapper) {
+        return new ElasticsearchTemplate(client, new CustomEntityMapper(objectMapper) );
+    }
+
+    public class CustomEntityMapper implements EntityMapper {
+
+        private ObjectMapper objectMapper;
+
+        public CustomEntityMapper(ObjectMapper objectMapper) {
+            this.objectMapper = objectMapper;
+            objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+            objectMapper.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+        }
+
+        @Override
+        public String mapToString(Object object) throws IOException {
+            return objectMapper.writeValueAsString(object);
+        }
+
+        @Override
+        public <T> T mapToObject(String source, Class<T> clazz) throws IOException {
+            return objectMapper.readValue(source, clazz);
+        }
+    }
+}

